### PR TITLE
[13.0][MIG] account_spread_cost_revenue

### PR DESCRIPTION
**Pull request by astirpe**

### CODE REVIEW

**Comment by astirpe** [comment]
Travis failure not related to this module:

```
account_asset_management/models/account_asset.py:294:51: E741 ambiguous variable name 'l'
account_asset_management/models/account_asset.py:654:35: E741 ambiguous variable name 'l'
account_asset_management/models/account_asset.py:1045:59: E741 ambiguous variable name 'l'
```

**Comment by astirpe** [comment]
@kittiu @casvissers would you review this PR?

**Comment by kittiu** [comment]
Got this error, when just create a new template
```
  File "/home/kittiu/odoo13/odoo/models.py", line 1747, in _add_missing_default_values
    defaults = self.default_get(list(missing_defaults))
  File "/home/kittiu/addons/account-financial-tools/account_spread_cost_revenue/models/account_spread_template.py", line 58, in default_get
    res["company_id"]
KeyError: 'company_id'
```

**Comment by kittiu** [comment]
I tested with case normal, it works with some concern,

Selection_061

The new invoice use line account = "Expense", but after create the spread table, it directily change invoice line account = "Prepayment" (as indicated in spread account).
This one will be ok.

BUT,
If I copy this invoice and set the same spread account = "Prepayment" in the wizard, it end up account "Prepayment" and "Prepayment"
image

This is more true, if non-account user who don't see the accounting column in invoice. The thing is, invoice was changed unnoticed.

I understand how this works, and seem the only way. But if it happen, the user can have hard time finding what happen.

May if the Spread Table is create with same Debit and Credit account, the system should throw an error message, explain what might cause?

**Comment by astirpe** [comment]
> May if the Spread Table is create with same Debit and Credit account, the system should throw an error message, explain what might cause?

Yes, I perfectly agree!

**Comment by kittiu** [comment]
Would you consider include enhancement from here once it is ready, thanks!
https://github.com/OCA/account-financial-tools/pull/993

**Comment by CasVissers-360ERP** [comment]
@astirpe 
Will the fix also be applied to v12?

**Comment by astirpe** [comment]
@CasVissers yes, of course. I will notify you when ready.

**Comment by astirpe** [comment]
~Tests are now failing due to https://github.com/odoo/odoo/pull/55088~
Tests fixed according to latest changes on Odoo 13:0: https://github.com/odoo/odoo/pull/55088

**Comment by kittiu** [comment]
@astirpe is this ready for review?

**Comment by astirpe** [comment]
@kittiu Yes it's ready for review

**Comment by kittiu** [comment]
> @kittiu Yes it's ready for review

Do you plan to include this change https://github.com/OCA/account-financial-tools/pull/993 now as another commit? Or this is just a plain MIG first?

**Comment by astirpe** [comment]
Right. I will include it next week.

**Comment by kittiu** [comment]
> Right. I will include it next week.

Let me know when this is ready, I will review.

**Comment by astirpe** [comment]
@kittiu done, would you give it a try?

File: account_spread_cost_revenue/readme/HISTORY.rst [hunk 1]
```diff
@@ -0,0 +1,23 @@
+12.0.2.0.0
```
Comment (kittiu) [inline account_spread_cost_revenue/readme/HISTORY.rst:6]: Would you also add 13.0.1.0.0 into history too.

**Comment by kittiu** [comment]
Steps to Error

- Create new database with demo data / no country
- Install module
- Create spread template
- Create vendor bill and try to create spread cost -> Error no currency_id when try to create account.spread.


Selection_201
image
image


Note:
====
If use auto spread like this, it works fine. The spread table is created correctly for this invoice.

image

**Comment by astirpe** [comment]
@kittiu thanks! Should be fixed now.

**Review by kittiu — APPROVED** [review]

**Comment by JordiBForgeFlow** [comment]
@adriagforgeflow can you review this one?

**Review by CasVissers-360ERP — APPROVED** [review]
We have this module in production for some time already..

**Comment by JordiBForgeFlow** [comment]
@astirpe if you do not remember to create the spread in the vendor bill the system won't stop you. Wouldn't it make more sense that some accounts require you to create a spread?

**Comment by astirpe** [comment]
@JordiBForgeFlow This module is used since years by our customers but none of them asked that requirement. I think that it could be a useful improvement, I'm curious what also other people think. In case there's enough interest I can develop it, unless you or your team want to contribute directly :)

**Comment by kittiu** [comment]
> @astirpe if you do not remember to create the spread in the vendor bill the system won't stop you. Wouldn't it make more sense that some accounts require you to create a spread?

You mean auto_spread? Is't based on template is enough?

https://github.com/onesteinbv/account-financial-tools/blob/13_mig_account_spread_cost_revenue/account_spread_cost_revenue/models/account_spread_template.py#L53

File: account_spread_cost_revenue/models/account_spread.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+    @api.depends(
+        "estimated_amount",
+        "currency_id",
+        "company_id",
+        "invoice_line_id.price_subtotal",
+        "invoice_line_id.currency_id",
+        "line_ids.amount",
+        "line_ids.move_id.state",
+    )
+    def _compute_amounts(self):
```
Comment (CasVissers-360ERP) [inline account_spread_cost_revenue/models/account_spread.py:166]: @astirpe This method does not take into account if the spread is for a credit or a normal invoice. Hence the amounts are calculated incorrectly for i.e. credit notes.

Comment (astirpe) [reply account_spread_cost_revenue/models/account_spread.py:166]: Thanks @CasVissers-360ERP. Do you have any solution already implemented for this?

**Comment by pedrobaeza** [comment]
@dreispt the bot didn't merge it. Do you launch it again?

**Comment by kittiu** [comment]
@astirpe there are error https://app.travis-ci.com/github/OCA/account-financial-tools/jobs/540709264#L2126
Can you rebase and/or fix this please. Thanks!

**Comment by astirpe** [comment]
@kittiu done, thanks!

**Pull request merged**